src/com/google/javascript/jscomp/ProductionCoverageInstrumentationCallback.java

```diff
@@ -137,56 +137,123 @@ private Node newInstrumentationNode(NodeTraversal traversal, Node node, String f
+   * can later be saved as a file.
    */
   private static final class ParameterMapping {
 
-    private final List<String> uniqueIdentifier;
-    private final List<String> paramValue;
+    // Values are stored as a mapping of the String of the encoded array indices to a String of the
+    // encoded uniqueIdentifier. This is so we can check if an encoded param has already been
+    // defined so that we do not create a duplicate. (Ex. Key: ACA (Base64 VLQ encoding of [0,1,0]),
+    // Value: C (Base64 VLQ encoding of the uniqueIdentifier). This map will later be inversed so
+    // that it is printed in the following form: C:ACA.
+    private final Map<String, String> paramValueEncodings;
+
+    // Values are stored as a mapping of String to Integers so that we can lookup the index of the
+    // encoded (file|function|type) name and also check if it is present in constant time. These
+    // mappings are added to the VariableMap once instrumentation is complete.
+    private final Map<String, Integer> fileNamesMap;
+    private final Map<String, Integer> functionNamesMap;
+    private final Map<String, Integer> typeMap;
+    private int fileNamesCounter;
+    private int functionNamesCounter;
+    private int typeCounter;
+
     private long nextUniqueIdentifier;
 
     ParameterMapping() {
       nextUniqueIdentifier = 0;
-      uniqueIdentifier = new ArrayList<>();
-      paramValue = new ArrayList<>();
+
+      paramValueEncodings = new HashMap<>();
+
+      // A LinkedHashMap is used so that when keys are printed, keySet() will obtain them in the
+      // insertion order which corroborates to the index. This helps to avoid the need of sorting
+      // by the Integer value and is more convenient.
+      fileNamesMap = new LinkedHashMap<>();
+      fileNamesCounter = 0;
+      functionNamesMap = new LinkedHashMap<>();
+      functionNamesCounter = 0;
+      typeMap = new LinkedHashMap<>();
+      typeCounter = 0;
     }
 
-    public String getEncodedParam(String fileName, String functionName, String type) {
-      String combinedParam = lenientFormat("%s %s %s", fileName, functionName, type);
-      long uniqueIdentifier = getUniqueIdentifier();
+    private String getEncodedParam(String fileName, String functionName, String type) {
 
-      if (uniqueIdentifier > Integer.MAX_VALUE) {
-        throw new ArithmeticException(
-            "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64"
-                + " VLQ");
+      if (!fileNamesMap.containsKey(fileName)) {
+        fileNamesMap.put(fileName, fileNamesCounter);
+        fileNamesCounter++;
+      }
+      if (!functionNamesMap.containsKey(functionName)) {
+        functionNamesMap.put(functionName, functionNamesCounter);
+        functionNamesCounter++;
       }
+      if (!typeMap.containsKey(type)) {
+        typeMap.put(type, typeCounter);
+        typeCounter++;
+      }
+
+      Integer fnNameIndex = functionNamesMap.get(functionName);
```

Comment:
How about declaring all of these within the try/catch?

General good practice is to limit the scope of variables as much as possible to the places where they are used

Reply:
I inlined the variables instead.